### ISSUE TITLE
Simplified API by using shared pointers - [merged]

### ISSUE THREAD

**Issue by wxmerkt** [loco-3d/crocoddyl#583]
In GitLab by @cmastalli on Oct 11, 2019, 23:33

_Merges topic/refact-using-shared-ptr -> devel_

Writing a complex optimal control problem can be very hard if the user needs to handle carefully all raw pointers. Paying this expensive lunch for just a bit of memory overhead (counter pointer) of shared pointers cannot be justified.

With this code, I ensure that we don't create runtime overhead.

Additionally, I simplified the unittest by using shared pointers. Please have a look of this code @MaximilienNaveau.

Finally, I have figured out how to expose custom data in Python. I included all these datas.

In conclusion this PR includes:
 1. Simpler API which uses shared pointers for abstracts (i.e. state, contacts, impulses, costs, actions, callbacks and shooting problem)
 2. benchmarks and unittest according 1.
 3. unittest code which uses shared pointers, removed custom management of memory
 4. exposition of custom data structures in Python

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 11, 2019, 23:53

added 1 commit

<ul><li>761debdf - [readme] Removed old dependencies + c++98 compliant!</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/245/diffs?diff_id=1255&start_sha=057fbb6b8f7c9fdea878057ab7f99742349126d1)

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 12, 2019, 07:17

merged

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 12, 2019, 07:17

mentioned in commit ec956830c7a73db8408e646a7804e67478482abe

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 12, 2019, 16:13

This pull request results in conflicts with other open PRs and breaks the Python bindings. We also don't have numbers on how much it affects performance (likely minimal) - which usually is a worry for other PRs. 

[I wasn't aware that an API restructuring was planned or mentioned during the last meeting(s)]

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 12, 2019, 16:41

I apologise for that. This prompt effort is to improve c++ API before everyone start to work on it.

It also helps @MaximilienNaveau to write quickly unittest code.

As I said before, I'll need to include few modifications in the API and merged this fixed error regarding the Python bindings of callbacks.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 13, 2019, 19:15

mentioned in merge request !246
